backend/repository/endpoint_update.go

```diff
@@ -69,6 +69,116 @@ func (r *Repository) WriteEndpointConfig(
+	if err != nil {
+		return err
+	}
+
+	// Remove the individual endpoint config json
+	if err = os.Remove(endpointConfigFile); err != nil {
+		return errors.Wrap(err, "failed to delete individual endpoint config json")
+	}
+
+	if len(groupConfig.Config.Endpoints) == 1 {
+		// this is the last endpoint remaining in this group, delete the group and update service config
+		if err = os.RemoveAll(endpointDir); err != nil {
+			return errors.Wrapf(err, "failed to remove endpoint group config directory for %s", endpointConfig.ID)
+		}
+
+		serviceConfigFile := filepath.Join(r.absPath(endpointCfgDir), "..", "/services/", gatewayCfg.ID, serviceConfigFileName)
+		serviceConfig := &codegen.EndpointClassConfig{}
+		if err = readJSONFile(serviceConfigFile, serviceConfig); err != nil {
+			return errors.Wrap(err, "failed to read service config json")
+		}
+		if i := findString(endpointConfig.ID, serviceConfig.Dependencies["endpoint"]); i != -1 {
+			serviceConfig.Dependencies["endpoint"] = append(serviceConfig.Dependencies["endpoint"][:i], serviceConfig.Dependencies["endpoint"][i+1:]...)
+		}
+		if err = writeToJSONFile(serviceConfigFile, serviceConfig); err != nil {
+			return errors.Wrap(err, "failed to write updated service config json")
+		}
+
+	} else {
+		// there are other endpoints in this group, need to update group config
+
+		// Remove this endpoint from the group list
+		groupConfig.Config.Endpoints = append(groupConfig.Config.Endpoints[:endpointIndexInGroup], groupConfig.Config.Endpoints[endpointIndexInGroup+1:]...)
+
+		// Remove client dependency from group config if no other endpoints in this group use it
+		otherEndpoints, _ := clientDep[endpointConfig.ClientID]
+		if len(filterStringsByPrefix(endpointConfig.ID+".", otherEndpoints)) == 1 {
+			if i := findString(endpointConfig.ClientID, groupConfig.Dependencies["client"]); i != -1 {
+				groupConfig.Dependencies["client"] = append(groupConfig.Dependencies["client"][:i], groupConfig.Dependencies["client"][i+1:]...)
+			}
+		}
+		// Remove middleware dependency from group config if no other endpoints in this group use it
+		for _, m := range endpointConfig.Middlewares {
+			otherEndpoints = getEndpointsWithMiddleware(gatewayCfg, m.Name, endpointConfig.ID)
+			if len(otherEndpoints) == 1 {
+				if i := findString(m.Name, groupConfig.Dependencies["middleware"]); i != -1 {
+					groupConfig.Dependencies["middleware"] = append(groupConfig.Dependencies["middleware"][:i], groupConfig.Dependencies["middleware"][i+1:]...)
+				}
+			}
+		}
+
+		if err = writeToJSONFile(endpointGroupConfigFile, groupConfig); err != nil {
+			return errors.Wrapf(err, "failed to write updated endpoint group config for %s", endpointConfig.ID)
+		}
+	}
+
+	return nil
+}
+
+func getEndpointsWithMiddleware(gatewayCfg *Config, middleware string, endpointGroup string) []string {
+	endpoints := make([]string, 0, 2)
+	for _, endpoint := range gatewayCfg.Endpoints {
+		if endpoint.ID == endpointGroup {
```

Comment:
why not just use a map here? You can still `len(map)`

Reply:
I guess in this case there's not much point to having a key, since the returned list is already limited to an endpoint group, and it can't be usefully stored since it might change when an endpoint is deleted